### ISSUE TITLE
How to redirect "sitename.com/any/path" to "www.sitename.com/any/path"?

### ISSUE THREAD

**Issue by mingyuguo** [gin-gonic/gin#646]
Basically I want to add the prefix www even if the users are not using it.

Maybe I can check context->request->url->host, and then redirect if I don't see www, but this requires that I do it for every handler.

Is there a better way, thanks!

**Comment by javierprovecho** [comment]
@mingyuguo you are right on checking the host through the context, but you can write a middleware that will be executed on every request.

Keep it simple:

```go
r.Use(func (c *gin.Context) {
    if !strings.HasPrefix(c.Request.Host, "www") {
        host := "www." + c.Request.Host
        c.Redirect(307, host+c.RequestURI)
    }
})
```

I just scratched this code from my smartphone. I hope you get the idea 😏.

**Comment by mingyuguo** [comment]
Greatly thanks javierprovecho for the amazing detailed answer!

For those who want a copy-paste solution:

I fixed two typos. For example, merely adding "www" may lead to: 

localhost -> localhost/www/localhost -> localhost/www/localhost/www/localhost -> infinite loop

So I changed to adding "http://www" instead.
Here is my final code (hope it doesn't break things)

I actually don't understand why we use c.Request.RequestURI instead of c.Request.URL (they seem to be always the same though based on my log)

router.Use(func(c *gin.Context) {
		if !strings.HasPrefix(c.Request.Host, "www") {
			if c.Request.Host != "127.0.0.1" {
				c.Redirect(307, "http://www."+c.Request.Host+c.Request.RequestURI)
			}
		}
	})
